httprequest_lego_provider/dns.py

```diff
@@ -14,21 +14,29 @@
 from .settings import GIT_REPO_URL
 
 FILENAME_TEMPLATE = "{domain}.domain"
-SPLIT_GIT_REPO_URL = GIT_REPO_URL.split("@")
-REPOSITORY_BASE_URL = "@".join(SPLIT_GIT_REPO_URL[:2])
-REPOSITORY_USER = (
-    SPLIT_GIT_REPO_URL[0].split("//")[1]
-    if SPLIT_GIT_REPO_URL and len(SPLIT_GIT_REPO_URL[0].split("//")) > 2
-    else None
-)
-REPOSITORY_BRANCH = SPLIT_GIT_REPO_URL[2] if len(SPLIT_GIT_REPO_URL) > 2 else None
 RECORD_CONTENT = "{record} 600 IN TXT \042{value}\042\n"
 
 
 class DnsSourceUpdateError(Exception):
     """Exception for DNS update errors."""
 
 
+def _parse_repository_url(repository_url: str) -> Tuple[str | None, str, str | None]:
+    """Get the parsed connection details from the repository connection string.
+
+    Args:
+        repository_url: the repository's connection string.
+
+    Returns:
+        the repository user.
+    """
+    splitted_url = repository_url.split("@")
+    user = splitted_url[0].split("//")[1]
+    base_url = "@".join(splitted_url[:2])
+    branch = splitted_url[2]
+    return user, base_url, branch
```

Comment:
Could we have a specific test for this function with different inputs?

Reply:
I was planning to add a regex on the charm to validate the config instead

Reply:
If this is always going to be a URL like this, I think using the `urllib.parse` library might be a little bit simpler to extract the username.

```python
import urllib.parse

def _parse_repository_url(repository_url):
    parsed = urllib.parse.urlparse(url)
    return parsed.username, *url.rsplit("@", 1)
```

Reply:
Not sure it's worth it just for one line, tbh
```
urllib.parse.urlparse("git+ssh://user@git.server/repo_name@lego").netloc.rsplit("@", 1)[0]
```
instead of 
```
user = splitted_url[0].split("//")[1]
```
since we'll need `splitted_url` for the rest of the fields

Reply:
I think this function can be compressed to just one line, but I agree, if there are some future requirements that need more parsing, this might be not applicable, up to you :)

```python
def _parse_repository_url(repository_url):
    return urllib.parse.urlparse(repository_url).username, *repository_url.rsplit("@", 1)
```